examples/games/mafia_werewolf/play_mafia.qnt

```diff
@@ -0,0 +1,185 @@
+    // Types for roles and life states of players
+    type Role = Mafia | Citizen
+    type LifeState = Alive | Dead
+    type Phase = Day | Night
+    type PlayerFeatures = {
+        role: Role,
+        status: LifeState,
+        voted: bool  // Indicates whether the player has voted
+    }
+    var player_to_features: str -> PlayerFeatures
+    var votes_by_player: str -> int
+    var game_phase: Phase
+    type Status = Pending | Done(Role)
+    var game_status: Status
+    /// Check if all alive players have voted
+    val all_voted = PLAYERS.filter(p => player_to_features.get(p).status == Alive)
+                      .forall(p => player_to_features.get(p).voted == true)
+    /// Check if there are any Mafia players left
+    val has_mafia = PLAYERS.exists(p => player_to_features.get(p).role == Mafia)
+    /// Check if there are any Citizen players left
+    val has_citizen = PLAYERS.exists(p => player_to_features.get(p).role == Citizen)
+    /// Find players with the most votes if all players have voted
+    val get_most_voted_players = {
+        if (all_voted) {
+            val max_votes = PLAYERS.fold(-1, (acc, p) => {
+                val votes = votes_by_player.get(p)
+                if (votes > acc) votes else acc
+            })
+            PLAYERS.filter(p => votes_by_player.get(p) == max_votes)
+        } else Set()  // Return an empty set if not all players have voted
+    }
+    /// Check if all Mafia players are dead
+    pure def all_mafias_dead(player: str -> PlayerFeatures): bool = {
+        PLAYERS.filter(p => player.get(p).role == Mafia).forall(p => player.get(p).status == Dead)
+    }
+    /// Check if all Citizen players are dead
+    pure def all_citizens_dead(player: str -> PlayerFeatures): bool = {
+        PLAYERS.filter(p => player.get(p).role == Citizen).forall(p => player.get(p).status == Dead)
+    }
+    /// Update the game status based on the current state
+    pure def update_status(player: str -> PlayerFeatures): Status = {
+        if (all_mafias_dead(player)) Done(Citizen)  // Citizens win if all Mafias are dead
+        else if (all_citizens_dead(player)) Done(Mafia)  // Mafia wins if all Citizens are dead
+        else Pending  // The game is still ongoing
+    }
+    /// Function to reset the players votes after voting is done
+    pure def reset_votes_and_statuses(player_to_features: str -> PlayerFeatures): str -> PlayerFeatures = {
+        PLAYERS.mapBy(p => {
+            ...player_to_features.get(p),
+            voted: false
+        })
+    }
+    /// Function to update player features after hanging
+    def update_features_after_hang(player_to_hang: str): str -> PlayerFeatures = {
+        PLAYERS.mapBy(p =>
+            if (p == player_to_hang) {
+                ...player_to_features.get(p),
+                status: Dead,
+                voted: false
+            } else {
+                ...player_to_features.get(p),
+                voted: false
+            }
+        )
```

Comment:
We can rewrite this using `setBy`. Also, I don't think we need to update `voted` here, since we don't account for this field when the player is dead anyway.
```suggestion
        player_to_features.setBy(player_to_hang, p => { ...p, status: Dead })
```